File: web/src/lib/api/devices/ring_device.ts
import requireResponseIsFromTrails from "$lib/api/requireResponseIsFromTrails";

export type RingDeviceResult =
    | { type: "success"; hasRingingStarted: boolean }
    | { type: "forbidden" }
    | { type: "timeout" }
    | { type: "error"; message: string };

async function postRing(deviceId: string, path: string): Promise<RingDeviceResult> {
    let response: Response;
    try {
        response = await fetch(`/api/v1/webapp/devices/${deviceId}/${path}`, { method: "POST" });
    } catch (e) {
        return { type: "error", message: e instanceof Error ? e.message : "Network error" };
    }
    requireResponseIsFromTrails(response);
    if (!response.ok) return { type: "error", message: `Request failed (${response.status})` };

    const data = await response.json();
    switch (data.type) {
        case "success":
            return { type: "success", hasRingingStarted: data.has_ringing_started ?? false };
        case "forbidden":
            return { type: "forbidden" };
        case "timeout":
            return { type: "timeout" };
        default:
            return { type: "error", message: data.message ?? "Unknown error" };
    }
}

/** Starts ringing one of the current user's own devices. */
export function ringDevice(deviceId: string): Promise<RingDeviceResult> {
    return postRing(deviceId, "ring");
}

/** Stops a ring previously started on one of the current user's own devices. */
export function stopRingDevice(deviceId: string): Promise<RingDeviceResult> {
    return postRing(deviceId, "ring/stop");
}

File: server/src/main/kotlin/es/jvbabi/trails/routes/webapp/devices/ringDevice.kt
package es.jvbabi.trails.routes.webapp.devices

import es.jvbabi.trails.api.TRAILS_WEBAPP_REALM
import es.jvbabi.trails.api.TrailsWebappPrincipal
import es.jvbabi.trails.data.DeviceSubscriptionMessage
import es.jvbabi.trails.data.DeviceSubscriptionRepository
import es.jvbabi.trails.database.DatabaseManager
import es.jvbabi.trails.database.Device
import es.jvbabi.trails.routes.app.deviceRingInfo
import es.jvbabi.trails.shared.dto.RingDeviceResponse
import io.ktor.server.application.*
import io.ktor.server.auth.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import org.koin.ktor.ext.inject
import kotlin.uuid.Uuid

/**
 * Starts ringing one of the current user's own devices. Like [webappPingDevice]
 * this is a standalone webapp request rather than a socket message, and reuses
 * the same device-subscription flow the app uses so the target device rings
 * regardless of whether the request came from the app or the web.
 */
fun Route.webappRingDevice() {
    val db by inject<DatabaseManager>()
    val deviceSubscriptionRepository by inject<DeviceSubscriptionRepository>()

    authenticate(TRAILS_WEBAPP_REALM) {
        post {
            val user = call.principal<TrailsWebappPrincipal>()!!.user
            val device = call.ownDeviceOrNull(db, user.id.value)
                ?: return@post call.respond<RingDeviceResponse>(RingDeviceResponse.Forbidden)

            deviceRingInfo[device.id.value] = user.username
            deviceSubscriptionRepository.getFlowForDeviceSubscription(device.id.value)
                .emit(DeviceSubscriptionMessage.Ring(device, pingedByDeviceName = user.username))

            call.respond<RingDeviceResponse>(RingDeviceResponse.Success(hasRingingStarted = true))
        }
    }
}

/**
 * Stops a ring previously started on one of the current user's own devices.
 */
fun Route.webappStopRingDevice() {
    val db by inject<DatabaseManager>()
    val deviceSubscriptionRepository by inject<DeviceSubscriptionRepository>()

    authenticate(TRAILS_WEBAPP_REALM) {
        post {
            val user = call.principal<TrailsWebappPrincipal>()!!.user
            val device = call.ownDeviceOrNull(db, user.id.value)
                ?: return@post call.respond<RingDeviceResponse>(RingDeviceResponse.Forbidden)

            deviceRingInfo.remove(device.id.value)
            deviceSubscriptionRepository.getFlowForDeviceSubscription(device.id.value)
                .emit(DeviceSubscriptionMessage.RingStop(device))

            call.respond<RingDeviceResponse>(RingDeviceResponse.Success(hasRingingStarted = false))
        }
    }
}

/**
 * Resolves the `{deviceId}` path parameter to a device owned by [userId], or
 * `null` if the id is missing/invalid or the device is not owned by the user.
 */
private suspend fun ApplicationCall.ownDeviceOrNull(db: DatabaseManager, userId: Uuid): Device? {
    val deviceId = parameters["deviceId"]?.let(Uuid::parseOrNull) ?: return null
    val device = db.transaction { Device.findById(deviceId) } ?: return null
    return if (db.transaction { device.owner.id.value == userId }) device else null
}
